Update changelog and snapshot docs (#8214)
hashicorp/nomad: website/pages/docs/commands/operator/snapshot-agent.mdx

---
layout: docs
page_title: 'Commands: operator snapshot agent'
sidebar_title: snapshot agent
description: |
  Periodically saves snapshots of Nomad server state
---

# Command: operator snapshot agent

> <EnterpriseAlert />

Author: This doesn't render for me locally but I noticed that consul uses it in for the banner in https://www.consul.io/docs/commands/snapshot/agent - will ask docs folks.

Reviewer: @angrycub had me do this:
```
~> **Enterprise Only!** This functionality only exists in Nomad
Enterprise. This is not present in the open source version of Nomad.
```

in https://github.com/hashicorp/nomad/blob/master/website/pages/docs/job-specification/multiregion.mdx

Reviewer: I think I see why: we don't have https://github.com/hashicorp/nomad/pull/7910 merged, which has that fancy new component to use. 😁

Reviewer 2: I *think* the reason Jeff didn't merge that was because he thought through a different abstraction which I implemented in my commit on this PR.

The agent can be run as a long-running daemon process or in a one-shot mode
from a batch job. As a long-running daemon, the agent will perform a leader
election so multiple processes can be run in a highly available fashion with
automatic failover. In daemon mode, the agent will also register itself with
Nomad as a service, along with health checks that show the agent is alive
and able to take snapshots.

If ACLs are enabled, a management token must be supplied in order to perform
snapshot operations.

The Config file has the following format (shown populated with default values):

```hcl
nomad {
  http_addr       = "127.0.0.1:8500"
  token           = ""
  region          = ""
  ca_file         = ""
  ca_path         = ""
  cert_file       = ""
  key_file        = ""
  tls_server_name = ""
}

snapshot {
  interval         = "1h"
  retain           = 30
  stale            = false
  service          = "nomad-snapshot"
  deregister_after = "72h"
  lock_key         = "nomad-snapshot/lock"
  max_failures     = 3
  name_prefix      = "nomad"
}

log {
  level           = "INFO"
  enable_syslog   = false
  syslog_facility = "LOCAL0"
}

> consul {

Reviewer 3: ```suggestion
nomad {
```

Author: This one is actually meant to be consul - nomad's block is above - agent can be configured to use consul for service discovery and locking.

  enabled         = true
  http_addr       = "127.0.0.1:8500"
  token           = ""
  datacenter      = ""
  ca_file         = ""
  ca_path         = ""
  cert_file       = ""
  key_file        = ""
  tls_server_name = ""
}

# one storage block is required

local_storage {
  path = "."
}

aws_storage {
  access_key_id     = ""
  secret_access_key = ""
  s3_region         = ""
  s3_endpoint       = ""
  s3_bucket         = ""
  s3_key_prefix     = "nomad-snapshot"
}

azure_blob_storage {
  account_name   = ""
  account_key    = ""
  container_name = ""
}

google_storage {
  bucket = ""
}
```

## Usage

```plaintext
nomad operator snapshot agent [options] <config_file>
```

## General Options

@include 'general_options.mdx'

## Snapshot agent Options

### Snapshot Options
- `-interval`: Interval at which to perform snapshots as a time with a unit suffix, which can be "s", "m", "h" for seconds, minutes, or hours. If 0 is provided, the agent will take a single snapshot and then exit, which is useful for running snapshots via batch jobs. Defaults to "1h".

> - `-lock-key`: A prefix in Consul's key-value store used to coordinate between different instances of the snapshot agent in order to only have one active instance at a time. For highly available operation of the snapshot agent, simply run multiple instances. All instances must be configured with the same lock key in order to properly coordinate. Defaults to "nomad-snapshot/lock".

Reviewer 3: ```suggestion
- `-lock-key`: A prefix in Nomad's key-value store used to coordinate between different instances of the snapshot agent in order to only have one active instance at a time. For highly available operation of the snapshot agent, simply run multiple instances. All instances must be configured with the same lock key in order to properly coordinate. Defaults to "nomad-snapshot/lock".
```

Author: Actually - upon reading this again - yes, this is meant to be Consul not Nomad :( will revert this change

- `-max-failures`: Number of snapshot failures after which the snapshot agent will give up leadership. In a highly available operation with multiple snapshot agents available, this gives another agent a chance to take over if an agent is experiencing issues, such as running out of disk space for snapshots. Defaults to 3.
- `-retain`: Number of snapshots to retain. After each snapshot is taken, the oldest snapshots will start to be deleted in order to retain at most this many snapshots. If this is set to 0, the agent will not perform this and snapshots will accumulate forever. Defaults to 30.

### Agent Options

> - `-deregister-after`: An interval, after which if the agent is unhealthy it will be automatically deregistered from Consul service. discovery. This is a time with a unit suffix, which can be "s", "m", "h" for seconds, minutes, or hours. If 0 is provided, this will be disabled. Defaults to "72h".

Reviewer 3: Maybe this one actually applies to consul?
```suggestion
- `-deregister-after`: An interval, after which if the agent is unhealthy it will be automatically deregistered from Nomad. discovery. This is a time with a unit suffix, which can be "s", "m", "h" for seconds, minutes, or hours. If 0 is provided, this will be disabled. Defaults to "72h".
```

Author: same here
